### PR TITLE
`schemadiff` small internal refactor: formalizing column charset/collation

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

This PR does some internal refactoring/cleanup without changing behavior. We give `ColumnDefinitionEntity` some more brainpower, cleanup column diffing (introducing `SetExplicitCharsetCollate` and simplifying `ColumnDiff()`.

## Related Issue(s)

As part of https://github.com/vitessio/vitess/issues/16229

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/schemadiff/table.go [hunk 1]
```diff
@@ -454,6 +454,15 @@ func NewCreateTableEntity(env *Environment, c *sqlparser.CreateTable) (*CreateTa
 	return entity, nil
 }
 
+func (c *CreateTableEntity) ColumnDefinitionEntities() []*ColumnDefinitionEntity {
```
Comment (rohit-nayak-ps) [inline go/vt/schemadiff/table.go:457]: I don't see where this is used ...

**Review by rohit-nayak-ps — APPROVED** [review]

File: go/vt/schemadiff/table.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/table.go:457]: You're right. This is preparation work for a later PR (WIP: https://github.com/vitessio/vitess/pull/16256).

**Review by rohit-nayak-ps — APPROVED** [review]

File: go/vt/schemadiff/column.go [hunk 2]
```diff
@@ -71,11 +71,69 @@ func NewModifyColumnDiffByDefinition(definition *sqlparser.ColumnDefinition) *Mo
...
+			return &UnknownColumnCollationCharsetError{Column: c.columnDefinition.Name.String(), Collation: c.columnDefinition.Type.Options.Collate}
+		}
+		c.columnDefinition.Type.Charset.Name = charset
+	}
+	if c.columnDefinition.Type.Charset.Name == "" {
+		// Still nothing? Assign the table's charset/collation.
+		c.columnDefinition.Type.Charset.Name = c.tableCharsetCollate.charset
+		if c.columnDefinition.Type.Options.Collate == "" {
+			c.columnDefinition.Type.Options.Collate = c.tableCharsetCollate.collate
+		}
```
Comment (mattlord) [inline go/vt/schemadiff/column.go:127]: Why do this when we blindly set `c.columnDefinition.Type.Options.Collate = c.tableCharsetCollate.collate` in the subsequent IF condition?

**Review by mattlord — APPROVED** [review]
Only one minor comment. There's no new tests added here, but I'm assuming that the behavior is covered by other tests, correct?

**Comment by shlomi-noach** [comment]
> There's no new tests added here, but I'm assuming that the behavior is covered by the existing tests, correct?

That's right. It's an internal refactor with no new behavioral change, hence covered by existing tests.

File: go/vt/schemadiff/column.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/column.go]: Good catch! Removed.

**Pull request merged**